core/src/engine.cpp

```diff
@@ -0,0 +1,55 @@
+#include <cstdlib>
+#include <exception>
+#include <vector>
+
+#include "core.h"
+#include "engine/model.h"
+#include "engine/openjtalk.h"
+#include "engine/synthesis_engine.h"
+
+#define NOT_INITIALIZED_OPENJTALK_ERR "Call initialize_openjtalk() first."
+
+using namespace voicevox::core::engine;
+
+static OpenJTalk *openjtalk = nullptr;
+static SynthesisEngine *engine = nullptr;
+
+bool initialize_openjtalk(const char *dict_path) {
+  // TODO: error handling
+  openjtalk = new OpenJTalk(dict_path);
+  return true;
+}
+
+
+uint8_t *voicevox_tts(const char *text, int64_t *speaker_id, int *binary_size) {
+  if (openjtalk == nullptr) {
+    throw std::runtime_error(NOT_INITIALIZED_OPENJTALK_ERR);
+  }
+  if (engine == nullptr) {
+    engine = new SynthesisEngine(openjtalk);
+  }
+
+  std::vector<AccentPhraseModel> accent_phrases = engine->create_accent_phrases(std::string(text), speaker_id);
+  accent_phrases = engine->replace_mora_data(accent_phrases, speaker_id);
+  const AudioQueryModel audio_query = {
+    accent_phrases,
+    1.0f,
+    0.0f,
+    1.0f,
+    1.0f,
+    0.1f,
+    0.1f,
+    engine->default_sampling_rate,
+    false,
+    "",
+  };
+
+  std::vector<unsigned char> wav = engine->synthesis_wave_format(audio_query, speaker_id, binary_size);
+  unsigned char *wav_heap = (unsigned char *)calloc(*binary_size, sizeof(unsigned char));
+  memcpy(wav_heap, wav.data(), sizeof(unsigned char) * (*binary_size));
```

Comment:
これは動作するコードです。
先程newは避けたほうが良いと指摘しましたが、この場合はnewしたほうが良いです。
newは避けるべきですが、C由来のmalloc,calloc,freeはもっと避けるべきです。(サイズ指定が煩雑で間違えやすいため)
以下のような感じに書けると思います。
```suggestion
const auto wav = engine->synthesis_wave_format(audio_query, speaker_id, binary_size);
auto* wav_heap = new uint8_t[*binary_size];
std::copy(wav.begin(),wav.end(),wav_heap);
```
考え方としては以下のように考えると良いです。
まずheap memoryを避け、stack memory上に載せられないか検討する
↓
heap memoryを使用することにした場合、vectorをメモリ領域代わりに使うか、C++のスマートポインタを使用できないか検討する
↓
Cなどの互換性のためvectorやスマートポインタを使えない場合heapメモリ確保をnew (配列)で行う。
↓
なんらかの事情でmallocやcalloc,freeを使わざるを得ない場合は使用する。

Reply:
なるほど、このようなやり方があるんですね...!(`new uint8_t`と言った記述は初めて知りました)
詳しい説明ありがとうございます、参考になります...!
suggestionのコードを利用させていただきます!